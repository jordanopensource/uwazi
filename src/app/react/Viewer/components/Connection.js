import PropTypes from 'prop-types';
import React, { Component } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { fromJS as Immutable } from 'immutable';
import { createSelector } from 'reselect';
import { Icon } from 'UI';
import { I18NLink } from 'app/I18N';
import { NeedAuthorization } from 'app/Auth';
import { withContext } from 'app/componentWrappers';
import ShowIf from 'app/App/ShowIf';
import { deleteReference } from 'app/Viewer/actions/referencesActions';
import {
  highlightReference,
  activateReference,
  selectReference,
  toggleReferences,
} from 'app/Viewer/actions/uiActions';
import { Item } from 'app/Layout';
import helpers from 'app/Documents/helpers';

const selectDoc = createSelector(
  s => s.documentViewer.targetDoc,
  s => s.documentViewer.doc,
  (targetDoc, doc) =>
    targetDoc.get('_id')
      ? helpers.performantDocToJSWithoutRelations(targetDoc)
      : helpers.performantDocToJSWithoutRelations(doc)
);

class Connection extends Component {
  clickReference(reference) {
    if (this.props.readOnly) {
      return;
    }

    if (!this.props.targetDoc) {
      this.props.activateReference(reference, this.props.referencesSection);
    }

    if (this.props.targetDoc && typeof reference.reference !== 'undefined') {
      this.props.selectReference(reference);
    }

    this.props.toggleReferences(true);
  }

  deleteReference(reference) {
    this.props.mainContext.confirm({
      accept: () => {
        this.props.deleteReference(reference);
      },
      title: 'Confirm delete connection',
      message: 'Are you sure you want to delete this connection?',
    });
  }

  relationType(id) {
    const type = this.props.relationTypes.find(relation => relation.get('_id') === id);
    if (type) {
      return type.name;
    }
  }

  render() {
    const { reference } = this.props;
    let itemClass = '';
    const disabled = this.props.targetDoc && typeof reference.reference === 'undefined';

    if (this.props.highlighted) {
      itemClass = 'relationship-hover';
    }

    if (this.props.active) {
      itemClass = 'relationship-active';
    }

    if (this.props.active && this.props.targetDoc && this.props.targetRange) {
      itemClass = 'relationship-selected';
    }

    if (this.props.readOnly) {
      itemClass = '';
    }

    const doc = Immutable(reference.associatedRelationship.entityData);
    return (
      <Item
        onMouseEnter={this.props.highlightReference.bind(null, reference._id)}
        onMouseLeave={this.props.highlightReference.bind(null, null)}
        onClick={this.clickReference.bind(this, reference)}
        doc={doc}
        noMetadata
        className={`${itemClass} item-${reference._id} ${disabled ? 'disabled' : ''} ${
          this.props.readOnly ? 'readOnly' : ''
        }`}
        data-id={reference._id}
        additionalText={
          reference.associatedRelationship.reference
            ? reference.associatedRelationship.reference.text
            : null
        }
        additionalMetadata={[
          { label: 'Connection type', value: this.relationType(reference.template) },
        ]}
        evalPublished
        buttons={
          <div className="item-shortcut-group">
            <ShowIf if={!this.props.targetDoc && !this.props.readOnly}>
              <NeedAuthorization roles={['admin', 'editor']}>
                <a
                  className="item-shortcut btn btn-default btn-hover-danger delete"
                  onClick={this.deleteReference.bind(this, reference)}
                >
                  <Icon icon="trash-alt" />
                </a>
              </NeedAuthorization>
            </ShowIf>

            <ShowIf if={!this.props.targetDoc}>
              <I18NLink
                to={`/entity/${doc.get('sharedId')}`}
                onClick={e => e.stopPropagation()}
                className="item-shortcut btn btn-default"
              >
                <Icon icon="file" />
              </I18NLink>
            </ShowIf>
          </div>
        }
      />
    );
  }
}

Connection.defaultProps = {
  targetDoc: false,
};

Connection.propTypes = {
  targetDoc: PropTypes.bool,
  readOnly: PropTypes.bool,
  highlighted: PropTypes.bool.isRequired,
  active: PropTypes.bool.isRequired,
  targetRange: PropTypes.object,
  relationTypes: PropTypes.object,
  reference: PropTypes.object.isRequired,
  highlightReference: PropTypes.func.isRequired,
  deleteReference: PropTypes.func,
  activateReference: PropTypes.func,
  selectReference: PropTypes.func,
  toggleReferences: PropTypes.func,
  referencesSection: PropTypes.string,
  mainContext: PropTypes.shape({
    confirm: PropTypes.func,
  }).isRequired,
};

const mapStateToProps = (state, ownProps) => {
  const { documentViewer } = state;
  const isActive = documentViewer.uiState.get('activeReference')
    ? documentViewer.uiState.get('activeReference') === ownProps.reference._id
    : documentViewer.uiState.get('activeReferences')?.includes(ownProps.reference._id);

  return {
    highlighted: documentViewer.uiState.get('highlightedReference') === ownProps.reference._id,
    active: isActive || false,
    targetRange: documentViewer.uiState.get('reference').get('targetRange'),
    targetDoc: !!documentViewer.targetDoc.get('_id'),
    relationTypes: documentViewer.relationTypes,
    doc: selectDoc(state),
  };
};

function mapDispatchToProps(dispatch) {
  return bindActionCreators(
    {
      highlightReference,
      activateReference,
      selectReference,
      deleteReference,
      toggleReferences,
    },
    dispatch
  );
}

export { Connection };
export default connect(mapStateToProps, mapDispatchToProps)(withContext(Connection));
