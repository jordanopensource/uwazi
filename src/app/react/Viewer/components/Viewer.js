/* eslint-disable max-lines */
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { Helmet } from 'react-helmet';
import PropTypes from 'prop-types';
import React, { Component } from 'react';
import { Map } from 'immutable';
import { ConnectionsList } from 'app/ConnectionsList';
import { CreateConnectionPanel } from 'app/Connections';
import { CurrentLocationLink, Icon } from 'app/Layout';
import { RelationshipsFormButtons } from 'app/Relationships';
import { Translate, I18NLink } from 'app/I18N';
import { actions } from 'app/BasicReducer';
import AddEntitiesPanel from 'app/Relationships/components/AddEntities';
import ContextMenu from 'app/ContextMenu';
import Footer from 'app/App/Footer';
import Marker from 'app/Viewer/utils/Marker';
import RelationshipMetadata from 'app/Relationships/components/RelationshipMetadata';
import ShowIf from 'app/App/ShowIf';
import { NeedAuthorization } from 'app/Auth';
import { FeatureToggle } from 'app/components/Elements/FeatureToggle';
import V2NewRelationshipsBoard from 'app/Entities/components/V2NewRelationshipsBoard';
import { PaginatorWithPage } from './Paginator';
import { addReference as addReferenceAction } from '../actions/referencesActions';
import {
  loadDefaultViewerMenu,
  loadTargetDocument as loadTargetDocumentAction,
} from '../actions/documentActions';
import { openPanel, toggleReferences } from '../actions/uiActions';
import { selectDoc } from '../selectors';
import ConfirmCloseForm from './ConfirmCloseForm';
import ViewMetadataPanel from './ViewMetadataPanel';
import ViewerDefaultMenu from './ViewerDefaultMenu';
import ViewerTextSelectedMenu from './ViewerTextSelectedMenu';
import SourceDocument from './SourceDocument';
import TargetDocument from './TargetDocument.js';
import determineDirection from '../utils/determineDirection';
import { OCRStatus } from './OCRStatus';

class Viewer extends Component {
  constructor(props) {
    super(props);
    this.state = { firstRender: true };
    this.handlePlainTextClick = this.handlePlainTextClick.bind(this);
  }

  componentDidMount() {
    const { store } = this.context;
    const { sidepanelTab } = this.props;
    store.dispatch(openPanel('viewMetadataPanel'));
    if (sidepanelTab === 'connections') {
      store.dispatch(actions.set('viewer.sidepanel.tab', ''));
    }
    store.dispatch(loadDefaultViewerMenu());
    Marker.init('div.main-wrapper');
    this.setState({ firstRender: false }); // eslint-disable-line react/no-did-mount-set-state
  }

  componentWillUnmount() {
    this.props.toggleReferences(true);
  }

  handlePlainTextClick() {
    const { showTab } = this.props;
    showTab('metadata');
  }

  prepareClassName(includeFooter) {
    const { panelIsOpen, targetDoc, showConnections } = this.props;
    let className = 'document-viewer with-header';
    className += panelIsOpen ? ' with-panel is-active' : '';
    className += targetDoc ? ' show-target-document' : '';
    className += showConnections ? ' connections' : '';
    className += includeFooter ? ' with-footer' : '';
    return className;
  }

  renderNoDoc() {
    const { doc } = this.props;
    return (
      <div className="row">
        <div className="content-header content-header-document">
          <div className="content-header-title">
            <Icon icon="lightbulb" />
            <Translate>
              This entity has no document, you probably want to see the metadata
            </Translate>
            &nbsp;
            <I18NLink to={`/entity/${doc.get('sharedId')}`}>
              <Translate>view</Translate>
            </I18NLink>
          </div>
        </div>
      </div>
    );
  }

  render() {
    const {
      doc,
      sidepanelTab,
      targetDoc,
      changePage,
      onPageChange,
      onDocumentReady,
      addReference,
      loadTargetDocument,
      panelIsOpen,
      showTextSelectMenu,
      file,
      user,
    } = this.props;
    const { firstRender } = this.state;
    if (doc.get('_id') && !doc.get('documents').size) {
      return this.renderNoDoc();
    }
    const includeFooter = user.get('_id') && sidepanelTab === 'relationships';
    const className = this.prepareClassName(includeFooter);
    const { raw, searchTerm, pageText, page } = this.props;
    const documentTitle = doc.get('title') ? doc.get('title') : '';

    return (
      <div className="row">
        <Helmet>
          <title>{`${documentTitle} â€¢ Page ${page}`}</title>
        </Helmet>
        <ShowIf if={!targetDoc}>
          <div className="content-header content-header-document">
            <div className="content-header-title">
              {sidepanelTab !== 'relationships' && (
                <>
                  <PaginatorWithPage totalPages={file.totalPages} onPageChange={changePage} />
                  <NeedAuthorization roles={['admin', 'editor']}>
                    <FeatureToggle feature="ocr.url">
                      <OCRStatus file={file} />
                    </FeatureToggle>
                  </NeedAuthorization>
                  <CurrentLocationLink
                    onClick={!raw ? this.handlePlainTextClick : () => {}}
                    className="btn btn-default"
                    queryParams={{ raw: raw || firstRender ? '' : 'true' }}
                  >
                    {raw || firstRender ? (
                      <Translate>Normal view</Translate>
                    ) : (
                      <Translate>Plain text</Translate>
                    )}
                  </CurrentLocationLink>
                </>
              )}
            </div>
          </div>
        </ShowIf>
        <main className={className}>
          <div className="main-wrapper">
            <ShowIf
              if={
                !['connections', 'relationships', 'newrelationships'].includes(sidepanelTab) &&
                !targetDoc
              }
            >
              {raw || firstRender ? (
                <div className={`${determineDirection(file)} raw-text`}>{pageText}</div>
              ) : (
                <SourceDocument
                  searchTerm={searchTerm}
                  onPageChange={onPageChange}
                  onDocumentReady={onDocumentReady}
                  file={file}
                />
              )}
            </ShowIf>
            <ShowIf if={sidepanelTab === 'connections' || sidepanelTab === 'relationships'}>
              <ConnectionsList hideFooter searchCentered />
            </ShowIf>
            <ShowIf if={this.props.newRelationshipsEnabled && sidepanelTab === 'newrelationships'}>
              <V2NewRelationshipsBoard sharedId={doc.get('sharedId')} />
            </ShowIf>
            <TargetDocument />
            <Footer />
          </div>
          {includeFooter && (
            <div className={`entity-footer remove-nesting ${panelIsOpen ? 'with-sidepanel' : ''}`}>
              <RelationshipsFormButtons />
            </div>
          )}
        </main>
        <ConfirmCloseForm />
        <ViewMetadataPanel
          raw={raw || firstRender}
          storeKey="documentViewer"
          searchTerm={searchTerm}
          file={file}
        />
        <CreateConnectionPanel
          containerId={targetDoc ? 'target' : doc.get('sharedId')}
          onCreate={addReference}
          onRangedConnect={loadTargetDocument}
          file={file}
        />
        {sidepanelTab === 'relationships' && (
          <>
            <RelationshipMetadata />
            <AddEntitiesPanel />
          </>
        )}
        <ContextMenu
          align={`bottom${includeFooter ? '-with-footer' : ''}`}
          overrideShow
          show={!panelIsOpen}
        >
          <ViewerDefaultMenu />
        </ContextMenu>
        <ContextMenu align="center" overrideShow show={showTextSelectMenu}>
          <ViewerTextSelectedMenu file={file} />
        </ContextMenu>
      </div>
    );
  }
}
Viewer.defaultProps = {
  searchTerm: '',
  raw: false,
  onPageChange: () => {},
  changePage: () => {},
  onDocumentReady: () => {},
  page: 1,
  doc: Map(),
  file: {},
  user: Map({}),
  // relationships v2
  newRelationshipsEnabled: false,
};
Viewer.propTypes = {
  searchTerm: PropTypes.string,
  raw: PropTypes.bool,
  onPageChange: PropTypes.func,
  changePage: PropTypes.func,
  onDocumentReady: PropTypes.func,
  doc: PropTypes.instanceOf(Map),
  pageText: PropTypes.string,
  panelIsOpen: PropTypes.bool,
  addReference: PropTypes.func,
  targetDoc: PropTypes.bool,
  sidepanelTab: PropTypes.string,
  loadTargetDocument: PropTypes.func,
  showConnections: PropTypes.bool,
  showTextSelectMenu: PropTypes.bool,
  selectedConnection: PropTypes.bool,
  selectedConnectionMetadata: PropTypes.object,
  showTab: PropTypes.func,
  page: PropTypes.number,
  locale: PropTypes.string.isRequired,
  file: PropTypes.object,
  user: PropTypes.instanceOf(Map),
  // relationships v2
  newRelationshipsEnabled: PropTypes.bool,
  toggleReferences: PropTypes.func,
};

Viewer.contextTypes = {
  store: PropTypes.object,
};

const mapStateToProps = state => {
  const { documentViewer } = state;
  const uiState = documentViewer.uiState.toJS();

  return {
    pageText: documentViewer.rawText,
    doc: selectDoc(state),
    panelIsOpen: !!uiState.panel,
    targetDoc: !!documentViewer.targetDoc.get('_id'),
    locale: state.locale,
    sidepanelTab: documentViewer.sidepanel.tab,
    showConnections: documentViewer.sidepanel.tab === 'references',
    showTextSelectMenu: Boolean(
      !documentViewer.targetDoc.get('_id') &&
        state.contextMenu.get('type') === 'ViewerTextSelectedMenu'
    ),
    user: state.user,
    // relationships v2
    newRelationshipsEnabled: state.settings?.collection?.get('features')?.get('newRelationships'),
  };
};

const mapDispatchToProps = dispatch =>
  bindActionCreators(
    {
      addReference: addReferenceAction,
      loadTargetDocument: loadTargetDocumentAction,
      showTab: tab => actions.set('viewer.sidepanel.tab', tab),
      toggleReferences,
    },
    dispatch
  );

const ConnectedViewer = connect(mapStateToProps, mapDispatchToProps)(Viewer);
export { ConnectedViewer };
