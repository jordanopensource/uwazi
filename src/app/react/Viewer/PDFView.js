import React, { Component } from 'react';
import { Helmet } from 'react-helmet';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import PropTypes from 'prop-types';
import { withRouter } from 'app/componentWrappers';
import { RequestParams } from 'app/utils/RequestParams';
import { actions } from 'app/BasicReducer';
import { isClient, events } from 'app/utils';
import { toUrlParams } from 'shared/JSONRequest';
import { ConnectedViewer as Viewer } from 'app/Viewer/components/Viewer';
import entitiesAPI from 'app/Entities/EntitiesAPI';
import { leaveEditMode } from 'app/Viewer/actions/documentActions';
import { searchParamsFromSearchParams } from 'app/utils/routeHelpers';
import { scrollToPage, activateReference } from './actions/uiActions';
import { requestViewerState } from './actions/routeActions';

const defaultDoc = entity => (entity.get('defaultDoc') ? entity.get('defaultDoc').toJS() : {});

class PDFViewComponent extends Component {
  static async requestState(requestParams, globalResources) {
    return requestViewerState(
      requestParams.add({ raw: requestParams.data.raw === 'true' || !isClient }),
      globalResources
    );
  }

  constructor(props, context) {
    super(props, context);
    this.changeBrowserHistoryPage = this.changeBrowserHistoryPage.bind(this);
    this.changePage = this.changePage.bind(this);
    this.onDocumentReady = this.onDocumentReady.bind(this);
  }

  componentDidMount() {
    const query = searchParamsFromSearchParams(this.props.searchParams);
    if (query.searchTerm) {
      this.context.store.dispatch(actions.set('viewer.sidepanel.tab', 'text-search'));
    }
  }

  componentWillReceiveProps(props) {
    const currentQuery = searchParamsFromSearchParams(this.props.searchParams);
    const query = searchParamsFromSearchParams(props.searchParams);

    if (query.page !== currentQuery.page && query.raw !== 'true') {
      this.changePage(query.page);
    }
    if (
      (query.page !== currentQuery.page || query.raw !== currentQuery.raw) &&
      query.raw === 'true'
    ) {
      entitiesAPI
        .getRawPage(new RequestParams({ _id: defaultDoc(props.entity)._id, page: query.page }))
        .then(pageText => {
          this.context.store.dispatch(actions.set('viewer/rawText', pageText));
        });
    }
  }

  componentWillUnmount() {
    this.props.leaveEditMode();
  }

  onDocumentReady(doc) {
    events.emit('documentLoaded');
    const query = searchParamsFromSearchParams(this.props.searchParams);

    if (query.raw === 'true') {
      return;
    }
    if (query.page) {
      scrollToPage(query.page, 0);
    }
    const { ref } = query;
    if (ref) {
      const reference = doc.get('relations').find(r => r.get('_id') === ref);
      this.context.store.dispatch(activateReference(reference.toJS()));
    }
  }

  changePage(nextPage) {
    const { raw = 'false' } = searchParamsFromSearchParams(this.props.searchParams);

    const notRaw = String(raw).toLowerCase() === 'false';
    if (notRaw) {
      return scrollToPage(nextPage);
    }

    return this.changeBrowserHistoryPage(nextPage, notRaw);
  }

  changeBrowserHistoryPage(newPage, replace = true) {
    const { page, ...queryWithoutPage } = searchParamsFromSearchParams(this.props.searchParams);

    this.props.navigate(
      `${this.props.location.pathname}${toUrlParams({ ...queryWithoutPage, page: newPage })}`,
      { replace }
    );
  }

  render() {
    const query = searchParamsFromSearchParams(this.props.searchParams);
    const page = Number(query.page || 1);
    const { filename } = defaultDoc(this.props.entity);
    const sharedId = this.props.entity.get('sharedId');
    const raw = query.raw === 'true' || !isClient;

    return (
      <>
        <Helmet>
          {raw && (
            <link rel="canonical" href={`/entity/${sharedId}?page=${page}&file=${filename}`} />
          )}
        </Helmet>
        <Viewer
          raw={raw}
          searchTerm={query.searchTerm}
          onPageChange={this.changeBrowserHistoryPage}
          onDocumentReady={this.onDocumentReady}
          changePage={this.changePage}
          page={page}
          file={defaultDoc(this.props.entity)}
        />
      </>
    );
  }
}

PDFViewComponent.contextTypes = {
  store: PropTypes.instanceOf(Object),
};

PDFViewComponent.propTypes = {
  entity: PropTypes.instanceOf(Object).isRequired,
  leaveEditMode: PropTypes.func,
  location: PropTypes.shape({
    pathname: PropTypes.string,
    search: PropTypes.string,
  }).isRequired,
  navigate: PropTypes.func.isRequired,
  searchParams: PropTypes.instanceOf(Object).isRequired,
};

PDFViewComponent.defaultProps = {
  leaveEditMode: () => {},
};

function mapDispatchToProps(dispatch) {
  return bindActionCreators({ leaveEditMode }, dispatch);
}

const SSRPDFView = connect(null, mapDispatchToProps)(withRouter(PDFViewComponent));
const PDFView = Object.assign(SSRPDFView, { requestState: PDFViewComponent.requestState });
export { PDFView, PDFViewComponent };
