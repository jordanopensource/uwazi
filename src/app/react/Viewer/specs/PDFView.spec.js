/**
 * @jest-environment jsdom
 */
/* eslint-disable max-statements */
/* eslint-disable react/no-multi-comp */
import React from 'react';
import { fromJS } from 'immutable';
import { shallow } from 'enzyme';
import entitiesAPI from 'app/Entities/EntitiesAPI';
import { actions } from 'app/BasicReducer';
import { PDFView, PDFViewComponent } from 'app/Viewer/PDFView';
import { ConnectedViewer as Viewer } from 'app/Viewer/components/Viewer';
import RouteHandler from 'app/App/RouteHandler';
import * as utils from 'app/utils';
import { RequestParams } from 'app/utils/RequestParams';
import { renderConnectedMount } from 'app/utils/test/renderConnected';
import * as documentActions from 'app/Viewer/actions/documentActions';
import * as routeActions from '../actions/routeActions';
import * as uiActions from '../actions/uiActions';

let page = 0;
let raw = 'abc';
let pathname = '';
let ref = '';
let anotherProp = '';

const mockUseLocation = jest.fn().mockImplementation(() => ({
  search: `?raw=${raw},page=${page}`,
  pathname,
}));

const mapProperties = props => {
  const map = new Map();
  Object.entries(props).forEach(([key, value]) => {
    map.set(key, value);
  });
  return map;
};
const mockUseSearchParams = jest.fn().mockImplementation(() => {
  const params = mapProperties({ raw, page });
  if (ref) {
    params.set('ref', ref);
  }
  if (anotherProp) {
    params.set('anotherProp', anotherProp);
  }

  return [params];
});
const mockNavigate = jest.fn().mockImplementation(path => path);
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useSearchParams: () => mockUseSearchParams(),
  useLocation: () => mockUseLocation(),
  useNavigate: () => path => mockNavigate(path),
  useMatches: () => {},
}));
jest.mock('app/ContextMenu', () => () => <div>ContextMenu</div>);
jest.mock('app/App/Footer', () => () => <div>Footer</div>);
jest.mock('app/Viewer/components/ViewMetadataPanel', () => () => <div>ViewMetadataPanel</div>);
jest.mock('app/Connections', () => ({
  CreateConnectionPanel: () => <div>CreateConnectionPanel</div>,
}));

describe('PDFView', () => {
  let component;
  let context;
  let props;

  const state = {
    documentViewer: {
      uiState: fromJS({ reference: { targetRange: [] } }),
      targetDoc: fromJS({ _id: 'document1' }),
      sidepanel: { tab: '1', metadata: { _id: 'prop1' } },
      targetDocReferences: [],
    },
    connections: { connection: fromJS({}) },
    user: fromJS({ _id: 'user1' }),
    settings: { collection: fromJS({}) },
    modals: fromJS({ ConfirmCloseForm: fromJS({ _id: 'document1' }) }),
    semanticSearch: { selectedDocument: fromJS({}) },
  };

  const render = () => {
    RouteHandler.renderedFromServer = true;
    component = renderConnectedMount(PDFView, state, props, true);
    component.instance().getChildContext().store.dispatch = context.store.dispatch;
  };

  beforeEach(() => {
    page = 1;
    raw = '';
    pathname = '';
    ref = '';
    anotherProp = '';

    const dispatch = jasmine.createSpy('dispatch');
    context = {
      store: {
        getState: () => ({}),
        dispatch: dispatch.and.callFake(action =>
          typeof action === 'function' ? action(dispatch) : action
        ),
        subscribe: jasmine.createSpy('subscribe'),
      },
    };

    props = {
      entity: fromJS({
        sharedId: 'a2b4c3',
        defaultDoc: { _id: 'documentId', sharedId: 'sharedId', filename: '1234.pdf' },
      }),
      routes: [],
    };
    spyOn(routeActions, 'requestViewerState');

    spyOn(routeActions, 'setViewerState').and.returnValue({ type: 'setViewerState' });
  });

  it('should pass down raw property', () => {
    raw = 'true';
    page = 2;
    render();
    expect(component.find(Viewer).props().raw).toEqual(true);
  });

  describe('when on server', () => {
    it('should always pass raw true', () => {
      raw = 'false';
      // eslint-disable-next-line no-import-assign
      utils.isClient = false;
      render();
      expect(component.find(Viewer).props().raw).toBe(true);
      // eslint-disable-next-line no-import-assign
      utils.isClient = true;
    });
  });

  it('should render the Viewer', () => {
    render();
    expect(component.find(Viewer).length).toBe(1);
  });

  describe('when raw', () => {
    it('should render link canonical to the not raw version', () => {
      raw = 'true';
      page = 1;
      pathname = 'pathname';
      render();

      expect(component.find({ link: [{ rel: 'canonical' }] })).toMatchSnapshot();
    });

    describe('when page is undefined', () => {
      it('should render link canonical with a default page', () => {
        raw = 'true';
        page = undefined;
        pathname = 'pathname';
        render();
        expect(component.find({ link: [{ rel: 'canonical' }] })).toMatchSnapshot();
      });
    });
  });
  describe('when not raw', () => {
    it('should not render link canonical', () => {
      raw = 'false';
      page = 1;
      pathname = 'pathname';
      render();
      expect(component.find({ link: [{ href: 'pathname?page=1', rel: 'canonical' }] }).length).toBe(
        0
      );
    });
  });
  describe('static requestState', () => {
    it('should call on requestViewerState', () => {
      const requestParams = new RequestParams({
        documentId: 'documentId',
        lang: 'es',
        page: 4,
        raw: 'true',
      });
      PDFView.requestState(requestParams, 'globalResources');
      expect(routeActions.requestViewerState).toHaveBeenCalledWith(
        new RequestParams({ documentId: 'documentId', lang: 'es', raw: true, page: 4 }),
        'globalResources'
      );
    });

    it('should modify raw to true if is server side rendered', () => {
      // eslint-disable-next-line no-import-assign
      utils.isClient = false;
      const requestParams = new RequestParams({
        documentId: 'documentId',
        lang: 'es',
        raw: 'false',
      });
      PDFView.requestState(requestParams, 'globalResources');
      expect(routeActions.requestViewerState).toHaveBeenCalledWith(
        new RequestParams({ documentId: 'documentId', lang: 'es', raw: true }),
        'globalResources'
      );
    });
  });

  describe('onDocumentReady', () => {
    it('should scrollToPage on the query when not on raw mode', () => {
      spyOn(uiActions, 'scrollToPage');
      raw = 'false';
      page = 15;
      render();
      component.find({ page: 15 }).at(0).props().onDocumentReady();
      expect(uiActions.scrollToPage).toHaveBeenCalledWith(15, 0);

      raw = 'true';
      render();
      uiActions.scrollToPage.calls.reset();
      component.find({ page: 15 }).at(0).props().onDocumentReady();
      expect(uiActions.scrollToPage).not.toHaveBeenCalled();
    });

    it('should activate text reference if query parameters have reference id', () => {
      spyOn(uiActions, 'activateReference').and.returnValue({ type: 'ABC' });
      raw = 'false';
      ref = 'refId';
      pathname = 'pathname';
      const reference = { _id: 'refId', range: { start: 200, end: 300 }, text: 'test' };
      const doc = fromJS({
        relations: [{ _id: 'otherRef' }, reference],
      });
      render();
      component.find({ page: 1 }).at(0).props().onDocumentReady(doc);
      expect(uiActions.activateReference).toHaveBeenCalledWith(reference);
    });

    it('should emit documentLoaded event', () => {
      spyOn(uiActions, 'scrollToPage');
      spyOn(utils.events, 'emit');
      render();

      component.find({ page: 1 }).at(0).props().onDocumentReady();
      expect(utils.events.emit).toHaveBeenCalledWith('documentLoaded');
    });
  });

  describe('changePage', () => {
    describe('when raw', () => {
      it('should changeBrowserHistoryPage', () => {
        raw = 'true';
        page = 15;
        anotherProp = 'test';
        pathname = 'pathname';
        spyOn(uiActions, 'scrollToPage');
        render();

        component.find({ page: 15 }).at(0).props().changePage(16);
        expect(mockNavigate).toHaveBeenCalledWith('pathname?raw=true&anotherProp=test&page=16');
        expect(uiActions.scrollToPage).not.toHaveBeenCalled();
      });
    });

    describe('when not raw', () => {
      it('should scrollToPage', () => {
        raw = 'false';
        page = 15;
        anotherProp = 'test';
        pathname = 'pathname';
        spyOn(uiActions, 'scrollToPage');
        mockNavigate.mockClear();

        render();
        component.find({ page: 15 }).at(0).props().changePage(16);
        expect(mockNavigate).not.toHaveBeenCalled();
        expect(uiActions.scrollToPage).toHaveBeenCalledWith(16);
      });
    });
  });

  const shallowComponent = searchParams =>
    shallow(
      <PDFViewComponent
        searchParams={searchParams}
        location={{ pathname: 'pathname' }}
        entity={fromJS({})}
        navigate={mockNavigate}
      />
    );

  describe('componentWillReceiveProps', () => {
    it('should load raw page when page/raw changes and raw is true', async () => {
      spyOn(entitiesAPI, 'getRawPage').and.returnValue(Promise.resolve('raw text'));
      const searchParams = mapProperties({ raw: 'true', page: 15 });
      const wrapper = shallowComponent(searchParams);
      expect(entitiesAPI.getRawPage).not.toHaveBeenCalled();
      wrapper.instance().context = context;
      mockNavigate.mockClear();
      entitiesAPI.getRawPage.calls.reset();
      searchParams.set('page', 16);
      searchParams.set('raw', 'false');
      wrapper.setProps({ searchParams });
      wrapper.update();
      expect(entitiesAPI.getRawPage).not.toHaveBeenCalled();
      entitiesAPI.getRawPage.calls.reset();
      const newSearchParams = mapProperties({ raw: 'true', page: 17 });

      wrapper.setProps({
        searchParams: newSearchParams,
        entity: fromJS({ defaultDoc: { _id: 'documentId' } }),
      });
      await wrapper.update();
      expect(context.store.dispatch).toHaveBeenCalledWith(
        actions.set('viewer/rawText', 'raw text')
      );
      expect(entitiesAPI.getRawPage).toHaveBeenCalledWith(
        new RequestParams({ _id: 'documentId', page: 17 })
      );
    });
  });

  describe('changeBrowserHistoryPage', () => {
    it('should push new browserHistory with new page', () => {
      const searchParams = mapProperties({ raw: 'true', page: 15, anotherProp: 'test' });
      const wrapper = shallowComponent(searchParams);
      wrapper.instance().changeBrowserHistoryPage(16);
      expect(mockNavigate).toHaveBeenCalledWith('pathname?raw=true&anotherProp=test&page=16', {
        replace: true,
      });

      mockNavigate.mockClear();
      searchParams.set('raw', 'false');
      searchParams.delete('anotherProp');
      wrapper.setProps({ searchParams });
      wrapper.update();
      wrapper.instance().changeBrowserHistoryPage(16);
      expect(mockNavigate).toHaveBeenCalledWith('pathname?raw=false&page=16', { replace: true });
    });
  });

  describe('componentWillUnmount', () => {
    it('should leave edit mode', () => {
      spyOn(documentActions, 'leaveEditMode').and.returnValue({ type: 'LEAVING_EDIT_MODE' });
      render();
      component.unmount();
      expect(documentActions.leaveEditMode).toHaveBeenCalled();
    });
  });
});
