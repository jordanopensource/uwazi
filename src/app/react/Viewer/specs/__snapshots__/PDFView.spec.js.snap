// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PDFView when raw should render link canonical to the not raw version 1`] = `
Array [
  <SideEffect(NullComponent)
    defer={true}
    encodeSpecialCharacters={true}
    link={
      Array [
        Object {
          "href": "/entity/a2b4c3?page=1&file=1234.pdf",
          "rel": "canonical",
        },
      ]
    }
  >
    <NullComponent
      defer={true}
      encodeSpecialCharacters={true}
      link={
        Array [
          Object {
            "href": "/entity/a2b4c3?page=1&file=1234.pdf",
            "rel": "canonical",
          },
        ]
      }
    />
  </SideEffect(NullComponent)>,
  <NullComponent
    defer={true}
    encodeSpecialCharacters={true}
    link={
      Array [
        Object {
          "href": "/entity/a2b4c3?page=1&file=1234.pdf",
          "rel": "canonical",
        },
      ]
    }
  />,
]
`;

exports[`PDFView when raw when page is undefined should render link canonical with a default page 1`] = `
Array [
  <SideEffect(NullComponent)
    defer={true}
    encodeSpecialCharacters={true}
    link={
      Array [
        Object {
          "href": "/entity/a2b4c3?page=1&file=1234.pdf",
          "rel": "canonical",
        },
      ]
    }
  >
    <NullComponent
      defer={true}
      encodeSpecialCharacters={true}
      link={
        Array [
          Object {
            "href": "/entity/a2b4c3?page=1&file=1234.pdf",
            "rel": "canonical",
          },
        ]
      }
    />
  </SideEffect(NullComponent)>,
  <NullComponent
    defer={true}
    encodeSpecialCharacters={true}
    link={
      Array [
        Object {
          "href": "/entity/a2b4c3?page=1&file=1234.pdf",
          "rel": "canonical",
        },
      ]
    }
  />,
]
`;
